File: LinkedList/206_ReverseLinkedList.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# @Author: xuezaigds@gmail.com


# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    # recursively reverse
    def reverseList(self, head):
        if not head or not head.next:
            return head
        reverse_head = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return reverse_head

    # iteratively reverse
    def reverseList_2(self, head):
        reverse_head = None
        while head:
            next_node = head.next
            head.next = reverse_head
            reverse_head = head
            head = next_node
        return reverse_head

"""
[]
[1]
[1,2]
[1,2,3,4,5]
"""

File: LinkedList/206_ReverseLinkedList.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-04-17 17:26:46
 */

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    // Recursively reverse
    ListNode* reverseList(ListNode* head) {
        if(head==NULL || head->next == NULL){
            return head;
        }
        ListNode *reversed_head = reverseList(head->next);
        head->next->next = head;
        head->next = NULL;
        return reversed_head;
    }
};

class Solution_2 {
public:
    // Itratively reverse
    ListNode* reverseList(ListNode* head) {
        ListNode *new_head = NULL;
        while(head!=NULL){
            ListNode *next_node = head->next;
            head->next = new_head;
            new_head = head;
            head = next_node;
        }
        return new_head;
    }
};

/*
[]
[1]
[1,2]
[1,2,3,4,5]
*/
